### ISSUE TITLE
UnboundLocalError in dyn.ext.ddhodge

### ISSUE THREAD

**Issue by lh12565** [aristoteleo/dynamo-release#289]
Hi, there is a error when I run ddhodge as below:
```python
dyn.ext.ddhodge(adata, basis='pca')
~/.local/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
  df_sub[k].cat.remove_unused_categories(inplace=True)
|-----------> not all cells are used, set diag to 1...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/.local/lib/python3.8/site-packages/dynamo/external/hodge.py", line 214, in ddhodge
    k = nbrs_idx.shape[1]
UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
```
I don't know how to solve it.

**Comment by Xiaojieqiu** [comment]
Sorry for this bug and thanks very much for using dynamo! I am busy with interviews and others. But @dummyindex can take a lead in addressing this question for you

**Comment by dummyindex** [comment]
Hi @lh12565, 

Thanks for using the package.
May I have your dynamo version (pip show dynamo) and the notebook/data you are running on?
If you used `pip install dynamo-release`, you can try installing the latest dynamo version
`
pip uninstall dynamo
pip install git+https://github.com/aristoteleo/dynamo-release
`

**Comment by lh12565** [comment]
> Hi @lh12565,
> 
> Thanks for using the package. May I have your dynamo version (pip show dynamo) and the notebook/data you are running on? If you used `pip install dynamo-release`, you can try installing the latest dynamo version `pip uninstall dynamo pip install git+https://github.com/aristoteleo/dynamo-release`

Hi, there is output of pip show dynamo:
```
Name: dynamo
Version: 0.0.0
Summary: Mapping Vector Field of Single Cells
Home-page: https://github.com/aristoteleo/dynamo-release
Author: Xiaojie Qiu, Yan Zhang, Ke Ni
Author-email: xqiu.sc@gmail.com
License: BSD
Location: /home/lh/.local/lib/python3.8/site-packages
Requires: anndata, colorcet, cvxopt, joblib, KDEpy, loompy, matplotlib, networkx, numba, numdifftools, numpy, pandas, PATSY, pynndescent, python-igraph, scikit-learn, scipy, seaborn, setuptools, statsmodels, tqdm, trimap, umap-learn
Required-by:
```
I installed the dynamo using:
```
git clone https://github.com/aristoteleo/dynamo-release.git
pip install dynamo-release/ --user
```
I run my own data which contain ~8000 cells.

**Comment by dummyindex** [comment]
> > Hi @lh12565,
> > Thanks for using the package. May I have your dynamo version (pip show dynamo) and the notebook/data you are running on? If you used `pip install dynamo-release`, you can try installing the latest dynamo version `pip uninstall dynamo pip install git+https://github.com/aristoteleo/dynamo-release`
> 
> Hi, there is output of pip show dynamo:
> 
> ```
> Name: dynamo
> Version: 0.0.0
> Summary: Mapping Vector Field of Single Cells
> Home-page: https://github.com/aristoteleo/dynamo-release
> Author: Xiaojie Qiu, Yan Zhang, Ke Ni
> Author-email: xqiu.sc@gmail.com
> License: BSD
> Location: /home/lh/.local/lib/python3.8/site-packages
> Requires: anndata, colorcet, cvxopt, joblib, KDEpy, loompy, matplotlib, networkx, numba, numdifftools, numpy, pandas, PATSY, pynndescent, python-igraph, scikit-learn, scipy, seaborn, setuptools, statsmodels, tqdm, trimap, umap-learn
> Required-by:
> ```
> 
> I installed the dynamo using:
> 
> ```
> git clone https://github.com/aristoteleo/dynamo-release.git
> pip install dynamo-release/ --user
> ```
> 
> I run my own data which contain ~8000 cells.

Hi @lh12565 ,

Thanks for letting us know this issue. 
I assume you are following this tutorial: https://dynamo-release.readthedocs.io/en/latest/notebooks/Differential_geometry.html?highlight=ddhodge
The data used in this differential geometry analysis is generated from this notebook: https://dynamo-release.readthedocs.io/en/latest/notebooks/zebrafish.html#

For the issue above, I believe you need to run preprocessing steps similar to what are presented in the link above. More specifically, `cell_velocities` generates `neighbors` information you need in `ddhodge`. You can try running these steps before calling ddhodge.

I agree with you that is a bug and we should handle it by either reminding users regarding it or running neighbors if the neighbor key is not present. (probably next week) For the time being, please try the solution above and let us know if it works.

Regards,
Dynamo team

**Comment by dummyindex** [comment]
> > Hi @lh12565,
> > Thanks for using the package. May I have your dynamo version (pip show dynamo) and the notebook/data you are running on? If you used `pip install dynamo-release`, you can try installing the latest dynamo version `pip uninstall dynamo pip install git+https://github.com/aristoteleo/dynamo-release`
> 
> Hi, there is output of pip show dynamo:
> 
> ```
> Name: dynamo
> Version: 0.0.0
> Summary: Mapping Vector Field of Single Cells
> Home-page: https://github.com/aristoteleo/dynamo-release
> Author: Xiaojie Qiu, Yan Zhang, Ke Ni
> Author-email: xqiu.sc@gmail.com
> License: BSD
> Location: /home/lh/.local/lib/python3.8/site-packages
> Requires: anndata, colorcet, cvxopt, joblib, KDEpy, loompy, matplotlib, networkx, numba, numdifftools, numpy, pandas, PATSY, pynndescent, python-igraph, scikit-learn, scipy, seaborn, setuptools, statsmodels, tqdm, trimap, umap-learn
> Required-by:
> ```
> 
> I installed the dynamo using:
> 
> ```
> git clone https://github.com/aristoteleo/dynamo-release.git
> pip install dynamo-release/ --user
> ```
> 
> I run my own data which contain ~8000 cells.

Hi @lh12565 ,

The issue is confirmed and we are fixing it. I will let you know when it is fixed and you can install again via pip (github ver).

Regards,
Dynamo team

**Comment by dummyindex** [comment]
Hi @lh12565 ,

The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?

Regards,
Dynamo team

**Comment by lh12565** [comment]
> Hi @lh12565 ,
> 
> The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> 
> Regards, Dynamo team

Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank.  
[note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)

```
In : adata=dyn.read('rep2_merge.h5ad')
In : adata
AnnData object with n_obs × n_vars = 8853 × 33033
    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
    var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
    obsm: 'X_umap'
    layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
In : adata.X.toarray()[0:5,0:5]
^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ],
       [0.        , 0.        , 0.        , 0.        , 0.        ]])
In : adata.obsm['X_umap']
array([[-2.14920756, -2.21941036],
       [ 2.41298774, -2.47172064],
       [ 3.23737934, -2.06542295],
       ...,
       [ 2.84649018, -3.82522577],
       [ 2.66472343, -2.02424306],
       [ 0.0903591 ,  1.25761837]])
```

**Comment by dummyindex** [comment]
> > Hi @lh12565 ,
> > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > Regards, Dynamo team
> 
> Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> 
> ```
> In : adata=dyn.read('rep2_merge.h5ad')
> In : adata
> AnnData object with n_obs × n_vars = 8853 × 33033
>     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
>     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
>     obsm: 'X_umap'
>     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> In : adata.X.toarray()[0:5,0:5]
> ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
>        [0.        , 0.        , 0.        , 0.        , 0.        ],
>        [0.        , 0.        , 0.        , 0.        , 0.        ],
>        [0.        , 0.        , 0.        , 0.        , 0.        ],
>        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> In : adata.obsm['X_umap']
> array([[-2.14920756, -2.21941036],
>        [ 2.41298774, -2.47172064],
>        [ 3.23737934, -2.06542295],
>        ...,
>        [ 2.84649018, -3.82522577],
>        [ 2.66472343, -2.02424306],
>        [ 0.0903591 ,  1.25761837]])
> ```

Thanks for sharing your code. I have reproduced your issue and fixed in the PR #290 and checked with zebrafish data.
Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values are the same.

To install the pending PR, you can install via 
```
pip uninstall dynamo
pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
```

**Comment by lh12565** [comment]
> > > Hi @lh12565 ,
> > > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > > Regards, Dynamo team
> > 
> > 
> > Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> > ```
> > In : adata=dyn.read('rep2_merge.h5ad')
> > In : adata
> > AnnData object with n_obs × n_vars = 8853 × 33033
> >     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
> >     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
> >     obsm: 'X_umap'
> >     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> > In : adata.X.toarray()[0:5,0:5]
> > ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
> >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> >        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> > In : adata.obsm['X_umap']
> > array([[-2.14920756, -2.21941036],
> >        [ 2.41298774, -2.47172064],
> >        [ 3.23737934, -2.06542295],
> >        ...,
> >        [ 2.84649018, -3.82522577],
> >        [ 2.66472343, -2.02424306],
> >        [ 0.0903591 ,  1.25761837]])
> > ```
> 
> Thanks for sharing your code. I have reproduced your issue, fixed in the PR #290 and checked with zebrafish data. Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values should be the same.
> 
> To try the pending PR (which should solve the problem), you can re-install dynamo via
> 
> ```
> pip uninstall dynamo
> pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
> ```
> 
> Please let me know if this fixes the issue on your side.

Hi, @dummyindex the error is still there. I didn't uninstall my old dynamo, but I install the new dynamo in new conda envs. I found they both installed in /home/~. I don't know if that's a mistake.

**Comment by dummyindex** [comment]
> > > > Hi @lh12565 ,
> > > > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > > > Regards, Dynamo team
> > > 
> > > 
> > > Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> > > ```
> > > In : adata=dyn.read('rep2_merge.h5ad')
> > > In : adata
> > > AnnData object with n_obs × n_vars = 8853 × 33033
> > >     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
> > >     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
> > >     obsm: 'X_umap'
> > >     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> > > In : adata.X.toarray()[0:5,0:5]
> > > ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
> > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > >        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> > > In : adata.obsm['X_umap']
> > > array([[-2.14920756, -2.21941036],
> > >        [ 2.41298774, -2.47172064],
> > >        [ 3.23737934, -2.06542295],
> > >        ...,
> > >        [ 2.84649018, -3.82522577],
> > >        [ 2.66472343, -2.02424306],
> > >        [ 0.0903591 ,  1.25761837]])
> > > ```
> > 
> > 
> > Thanks for sharing your code. I have reproduced your issue, fixed in the PR #290 and checked with zebrafish data. Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values should be the same.
> > To try the pending PR (which should solve the problem), you can re-install dynamo via
> > ```
> > pip uninstall dynamo
> > pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please let me know if this fixes the issue on your side.
> 
> Hi, @dummyindex the error is still there. I didn't uninstall my old dynamo, but I install the new dynamo in new conda envs. I found they both installed in /home/~. I don't know if that's a mistake.

For the new environment, did you `conda activate your_env` (or `source activate your_env` on certain cluster)? 
May I have your new error message? If the exception line number is exactly the same as your previous one, then the new version is not be installed correctly.

**Comment by lh12565** [comment]
> > > > > Hi @lh12565 ,
> > > > > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > > > > Regards, Dynamo team
> > > > 
> > > > 
> > > > Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> > > > ```
> > > > In : adata=dyn.read('rep2_merge.h5ad')
> > > > In : adata
> > > > AnnData object with n_obs × n_vars = 8853 × 33033
> > > >     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
> > > >     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
> > > >     obsm: 'X_umap'
> > > >     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> > > > In : adata.X.toarray()[0:5,0:5]
> > > > ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
> > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > >        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> > > > In : adata.obsm['X_umap']
> > > > array([[-2.14920756, -2.21941036],
> > > >        [ 2.41298774, -2.47172064],
> > > >        [ 3.23737934, -2.06542295],
> > > >        ...,
> > > >        [ 2.84649018, -3.82522577],
> > > >        [ 2.66472343, -2.02424306],
> > > >        [ 0.0903591 ,  1.25761837]])
> > > > ```
> > > 
> > > 
> > > Thanks for sharing your code. I have reproduced your issue, fixed in the PR #290 and checked with zebrafish data. Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values should be the same.
> > > To try the pending PR (which should solve the problem), you can re-install dynamo via
> > > ```
> > > pip uninstall dynamo
> > > pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Please let me know if this fixes the issue on your side.
> > 
> > 
> > Hi, @dummyindex the error is still there. I didn't uninstall my old dynamo, but I install the new dynamo in new conda envs. I found they both installed in /home/~. I don't know if that's a mistake.
> 
> Thanks for letting us know. For the new environment, did you `conda activate your_env` (or `source activate your_env` in certain HPC enviroments)? May I have your new error message? If the exception line number is exactly the same as your previous one, then the new version is not be installed correctly.

Hi, @dummyindex yes, I did source activate my_env before install new dynamo. The error is same as before. I will uninstall it and install the new one and run again.
```
/home/lh/.local/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
  df_sub[k].cat.remove_unused_categories(inplace=True)
|-----------> not all cells are used, set diag to 1...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/lh/.local/lib/python3.8/site-packages/dynamo_release-1.0.0-py3.8.egg/dynamo/external/hodge.py", line 214, in ddhodge
    k = nbrs_idx.shape[1]
UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
```

**Comment by dummyindex** [comment]
> > > > > > Hi @lh12565 ,
> > > > > > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > > > > > Regards, Dynamo team
> > > > > 
> > > > > 
> > > > > Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> > > > > ```
> > > > > In : adata=dyn.read('rep2_merge.h5ad')
> > > > > In : adata
> > > > > AnnData object with n_obs × n_vars = 8853 × 33033
> > > > >     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
> > > > >     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
> > > > >     obsm: 'X_umap'
> > > > >     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> > > > > In : adata.X.toarray()[0:5,0:5]
> > > > > ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
> > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> > > > > In : adata.obsm['X_umap']
> > > > > array([[-2.14920756, -2.21941036],
> > > > >        [ 2.41298774, -2.47172064],
> > > > >        [ 3.23737934, -2.06542295],
> > > > >        ...,
> > > > >        [ 2.84649018, -3.82522577],
> > > > >        [ 2.66472343, -2.02424306],
> > > > >        [ 0.0903591 ,  1.25761837]])
> > > > > ```
> > > > 
> > > > 
> > > > Thanks for sharing your code. I have reproduced your issue, fixed in the PR #290 and checked with zebrafish data. Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values should be the same.
> > > > To try the pending PR (which should solve the problem), you can re-install dynamo via
> > > > ```
> > > > pip uninstall dynamo
> > > > pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > Please let me know if this fixes the issue on your side.
> > > 
> > > 
> > > Hi, @dummyindex the error is still there. I didn't uninstall my old dynamo, but I install the new dynamo in new conda envs. I found they both installed in /home/~. I don't know if that's a mistake.
> > 
> > 
> > Thanks for letting us know. For the new environment, did you `conda activate your_env` (or `source activate your_env` in certain HPC enviroments)? May I have your new error message? If the exception line number is exactly the same as your previous one, then the new version is not be installed correctly.
> 
> Hi, @dummyindex yes, I did source activate my_env before install new dynamo. The error is same as before. I will uninstall it and install the new one and run again.
> 
> ```
> /home/lh/.local/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
>   df_sub[k].cat.remove_unused_categories(inplace=True)
> |-----------> not all cells are used, set diag to 1...
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/lh/.local/lib/python3.8/site-packages/dynamo_release-1.0.0-py3.8.egg/dynamo/external/hodge.py", line 214, in ddhodge
>     k = nbrs_idx.shape[1]
> UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
> ```

Your error messages indeed is from the older version. 
Sounds good. Please let us know if there is any issue with the PR #290 version regarding `ddhodge` in this thread.

**Comment by lh12565** [comment]
> > > > > > > Hi @lh12565 ,
> > > > > > > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > > > > > > Regards, Dynamo team
> > > > > > 
> > > > > > 
> > > > > > Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> > > > > > ```
> > > > > > In : adata=dyn.read('rep2_merge.h5ad')
> > > > > > In : adata
> > > > > > AnnData object with n_obs × n_vars = 8853 × 33033
> > > > > >     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
> > > > > >     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
> > > > > >     obsm: 'X_umap'
> > > > > >     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> > > > > > In : adata.X.toarray()[0:5,0:5]
> > > > > > ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
> > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> > > > > > In : adata.obsm['X_umap']
> > > > > > array([[-2.14920756, -2.21941036],
> > > > > >        [ 2.41298774, -2.47172064],
> > > > > >        [ 3.23737934, -2.06542295],
> > > > > >        ...,
> > > > > >        [ 2.84649018, -3.82522577],
> > > > > >        [ 2.66472343, -2.02424306],
> > > > > >        [ 0.0903591 ,  1.25761837]])
> > > > > > ```
> > > > > 
> > > > > 
> > > > > Thanks for sharing your code. I have reproduced your issue, fixed in the PR #290 and checked with zebrafish data. Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values should be the same.
> > > > > To try the pending PR (which should solve the problem), you can re-install dynamo via
> > > > > ```
> > > > > pip uninstall dynamo
> > > > > pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > Please let me know if this fixes the issue on your side.
> > > > 
> > > > 
> > > > Hi, @dummyindex the error is still there. I didn't uninstall my old dynamo, but I install the new dynamo in new conda envs. I found they both installed in /home/~. I don't know if that's a mistake.
> > > 
> > > 
> > > Thanks for letting us know. For the new environment, did you `conda activate your_env` (or `source activate your_env` in certain HPC enviroments)? May I have your new error message? If the exception line number is exactly the same as your previous one, then the new version is not be installed correctly.
> > 
> > 
> > Hi, @dummyindex yes, I did source activate my_env before install new dynamo. The error is same as before. I will uninstall it and install the new one and run again.
> > ```
> > /home/lh/.local/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
> >   df_sub[k].cat.remove_unused_categories(inplace=True)
> > |-----------> not all cells are used, set diag to 1...
> > Traceback (most recent call last):
> >   File "<stdin>", line 1, in <module>
> >   File "/home/lh/.local/lib/python3.8/site-packages/dynamo_release-1.0.0-py3.8.egg/dynamo/external/hodge.py", line 214, in ddhodge
> >     k = nbrs_idx.shape[1]
> > UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
> > ```
> 
> Your error messages indeed is from the older version. Sounds good. Please let us know if there is any issue with the PR #290 version regarding `ddhodge` in this thread.

Hi, @dummyindex I re-install the dynamo, but the error is still there ╮(╯▽╰)╭. 
```
/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
  df_sub[k].cat.remove_unused_categories(inplace=True)
|-----------> not all cells are used, set diag to 1...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/dynamo/external/hodge.py", line 214, in ddhodge
    k = nbrs_idx.shape[1]
UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
```

**Comment by dummyindex** [comment]
> > > > > > > > Hi @lh12565 ,
> > > > > > > > The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> > > > > > > > Regards, Dynamo team
> > > > > > > 
> > > > > > > 
> > > > > > > Thanks for your quick reply！There is my code in note.zip. The "rep2_merge.h5ad" merged by my loom file of velocity and the normlised seurat data, and it already have umap (no PCA?) and normlised matrix(no counts?) as below. By the way, my out put of full_reg_rank is same as full_eff_rank. [note.zip](https://github.com/aristoteleo/dynamo-release/files/8150208/note.zip)
> > > > > > > ```
> > > > > > > In : adata=dyn.read('rep2_merge.h5ad')
> > > > > > > In : adata
> > > > > > > AnnData object with n_obs × n_vars = 8853 × 33033
> > > > > > >     obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.ribo', 'percent.HB', 'integrated_snn_res.0.1', 'seurat_clusters', 'integrated_snn_res.1', 'celltype', 'copykat', 'integrated_snn_res.0.2', 'sample_batch', 'Clusters', '_X', '_Y', 'initial_size_spliced', 'initial_size_unspliced', 'initial_size'
> > > > > > >     var: 'features', 'Accession', 'Chromosome', 'End', 'Start', 'Strand'
> > > > > > >     obsm: 'X_umap'
> > > > > > >     layers: 'ambiguous', 'matrix', 'spliced', 'unspliced'
> > > > > > > In : adata.X.toarray()[0:5,0:5]
> > > > > > > ^[[Aarray([[0.        , 0.59050173, 0.        , 0.        , 0.        ],
> > > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ],
> > > > > > >        [0.        , 0.        , 0.        , 0.        , 0.        ]])
> > > > > > > In : adata.obsm['X_umap']
> > > > > > > array([[-2.14920756, -2.21941036],
> > > > > > >        [ 2.41298774, -2.47172064],
> > > > > > >        [ 3.23737934, -2.06542295],
> > > > > > >        ...,
> > > > > > >        [ 2.84649018, -3.82522577],
> > > > > > >        [ 2.66472343, -2.02424306],
> > > > > > >        [ 0.0903591 ,  1.25761837]])
> > > > > > > ```
> > > > > > 
> > > > > > 
> > > > > > Thanks for sharing your code. I have reproduced your issue, fixed in the PR #290 and checked with zebrafish data. Regarding your `full_reg` and `full_eff` issue, it is normal because you set "abs=True". Their absolute values should be the same.
> > > > > > To try the pending PR (which should solve the problem), you can re-install dynamo via
> > > > > > ```
> > > > > > pip uninstall dynamo
> > > > > > pip install git+https://github.com/dummyindex/dynamo-release.git@update-docs
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > Please let me know if this fixes the issue on your side.
> > > > > 
> > > > > 
> > > > > Hi, @dummyindex the error is still there. I didn't uninstall my old dynamo, but I install the new dynamo in new conda envs. I found they both installed in /home/~. I don't know if that's a mistake.
> > > > 
> > > > 
> > > > Thanks for letting us know. For the new environment, did you `conda activate your_env` (or `source activate your_env` in certain HPC enviroments)? May I have your new error message? If the exception line number is exactly the same as your previous one, then the new version is not be installed correctly.
> > > 
> > > 
> > > Hi, @dummyindex yes, I did source activate my_env before install new dynamo. The error is same as before. I will uninstall it and install the new one and run again.
> > > ```
> > > /home/lh/.local/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
> > >   df_sub[k].cat.remove_unused_categories(inplace=True)
> > > |-----------> not all cells are used, set diag to 1...
> > > Traceback (most recent call last):
> > >   File "<stdin>", line 1, in <module>
> > >   File "/home/lh/.local/lib/python3.8/site-packages/dynamo_release-1.0.0-py3.8.egg/dynamo/external/hodge.py", line 214, in ddhodge
> > >     k = nbrs_idx.shape[1]
> > > UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
> > > ```
> > 
> > 
> > Your error messages indeed is from the older version. Sounds good. Please let us know if there is any issue with the PR #290 version regarding `ddhodge` in this thread.
> 
> Hi, @dummyindex I re-install the dynamo, but the error is still there ╮(╯▽╰)╭.
> 
> ```
> /NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
>   df_sub[k].cat.remove_unused_categories(inplace=True)
> |-----------> not all cells are used, set diag to 1...
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/dynamo/external/hodge.py", line 214, in ddhodge
>     k = nbrs_idx.shape[1]
> UnboundLocalError: local variable 'nbrs_idx' referenced before assignment
> ```

The new version should have `k = nbrs_idx.shape[1]` on line 224. You can wait for the PR merge (should be Monday morning EST), or try installing the PR version again.

**Comment by lh12565** [comment]
> Hi @lh12565 ,
> 
> The issue should be fixed when #290 is merged into master branch. Meanwhile, may I have your notebook scripts for processing adata with dynamo to reproduce the issue and confirm the issue is fixed on my side?
> 
> Regards, Dynamo team

Hi, @dummyindex the issue has been solved, but there is another error:
```
/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/anndata/_core/anndata.py:1095: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.
  df_sub[k].cat.remove_unused_categories(inplace=True)
|-----------> not all cells are used, set diag to 1...
|-----------> Constructing W matrix according upsampling=True and downsampling=True options...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/dynamo/external/hodge.py", line 230, in ddhodge
    W = csr_matrix(
  File "/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/scipy/sparse/_compressed.py", line 53, in __init__
    self._coo_container(arg1, shape=shape, dtype=dtype)
  File "/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/scipy/sparse/_coo.py", line 196, in __init__
    self._check()
  File "/NAS/lh/software/miniconda3/envs/dynamonew/lib/python3.8/site-packages/scipy/sparse/_coo.py", line 283, in _check
    raise ValueError('row index exceeds matrix dimensions')
ValueError: row index exceeds matrix dimensions
```

**Comment by dummyindex** [comment]
Hi @lh12565 ,

Yes the issue was confirmed. We fixed it and it should work now. Please try
```
source activate dynamonew
pip uninstall dynamo
pip install git+https://github.com/dummyindex/dynamo-release.git@fix-issue-289
```
Or you can wait the #291 to be merged soon.
Sorry for glitches and thanks for your report. Please let us know if this fixes the problem.

**Comment by lh12565** [comment]
> Hi @lh12565 ,
> 
> Yes the issue was confirmed. We fixed it and it should work now. Please try
> 
> ```
> source activate dynamonew
> pip uninstall dynamo
> pip install git+https://github.com/dummyindex/dynamo-release.git@fix-issue-289
> ```
> 
> Or you can wait the #291 to be merged soon. Sorry for the glitches and thanks for your report. Please let us know if this fixes the problem.

Thanks! It works! You are the most efficient person I have ever met.
